torch_xla/csrc/status.h

```diff
@@ -0,0 +1,81 @@
+#ifndef XLA_TORCH_XLA_CSRC_STATUS_H_
+#define XLA_TORCH_XLA_CSRC_STATUS_H_
+
+#include "absl/status/statusor.h"
+
+namespace torch_xla {
+
+// Creates a new Status instance, appending the current location (e.g. file and
+// line information) to the status message.
+//
+// This should be used whenever we are returning new error status, instead of
+// propagating. Then, if `kEnvShowCppErrorContext` environment variable is set,
+// the location information will be shown.
+#define XLA_ERROR_WITH_LOCATION(STATUS) \
+  ::torch_xla::MaybeWithLocation(std::move(STATUS), __FILE__, __LINE__)
+
+#define _XLA_RETURN_IF_ERROR(EXPR, THEN, ...)                         \
+  do {                                                                \
+    auto _status_var = (EXPR);                                        \
+    if (!_status_var.ok()) {                                          \
+      return ::torch_xla::MaybeWithNewMessage(                        \
+          std::move(_status_var), __FILE__, __LINE__, ##__VA_ARGS__); \
+    }                                                                 \
+    THEN;                                                             \
+  } while (0)
+
+// Propagates `REXPR`, in case it's a non-ok status.
+#define XLA_RETURN_IF_ERROR(REXPR, ...) \
+  _XLA_RETURN_IF_ERROR(REXPR, {}, ##__VA_ARGS__)
+
+// Propagates `REXPR`, in case it's a non-ok status. Otherwise, assign
+// its result to `LHS`.
+#define XLA_ASSIGN_OR_RETURN(LHS, REXPR, ...) \
```

Comment:
TensorFlow and protobuf's `ASSIGN_OR_RETURN(T x, expr);` can define a new variable `T x` that can be used after the current statement. This is really useful. This implementation cannot do it as `LHS` goes out of scope after this macro.

Reply:
Modified it, so that we can use it that way, now.